src/libspu/mpc/shamir/arithmetic.cc

```diff
@@ -0,0 +1,434 @@
+        _r_t[idx * (world_size - th) + i] = ret_t(0, i).data();
+        _r_2t[idx * (world_size - th) + i] = ret_2t(0, i).data();
+      }
+    });
+  });
+  return out;
+}
+
+}  // namespace
+
+// Ref: DN'07 protocol for honesty majority
+//  https://www.iacr.org/archive/crypto2007/46220565/46220565.pdf
+// [Offline Phase]
+NdArrayRef RandA::proc(KernelEvalContext* ctx, const Shape& shape) const {
+  auto* prg_state = ctx->getState<PrgState>();
+  auto* comm = ctx->getState<Communicator>();
+  const auto field = ctx->getState<Z2kState>()->getDefaultField();
+  NdArrayRef out = ring_zeros(field, shape);
+  int64_t world_size = comm->getWorldSize();
+  int64_t th = ctx->sctx()->config().sss_threshold();
+  int64_t numel = shape.numel();
+
+  // run one-time DN protocol we can generate (world_size-th) random shares
+  auto dn_times = (numel - 1) / (world_size - th) + 1;
+  auto ty = makeType<GfmpTy>(field);
+  auto r = gfmp_mod(prg_state->genPriv(field, {dn_times}).as(ty));
+  auto shares_r = gfmp_rand_shamir_shares(r, world_size, th);
+  auto rank = comm->getRank();
+
+  std::vector<NdArrayRef> r_shrs(world_size);
+  for (size_t i = 0; i < shares_r.size(); ++i) {
+    if (i != rank) {
+      comm->sendAsync(i, shares_r[i], "send r_share");
+    }
+  }
+  for (size_t i = 0; i < shares_r.size(); ++i) {
+    if (i != rank) {
+      r_shrs[i] = comm->recv(i, shares_r[i].eltype(), "send r_share");
+    }
+  }
+  comm->addCommStatsManually(1, r.elsize() * r.numel() * (world_size - 1));
+
+  r_shrs[rank] = shares_r[rank];
+  return DISPATCH_ALL_FIELDS(field, [&]() {
+    NdArrayRef r_t(ty, {dn_times * (world_size - th)});
+    NdArrayView<ring2k_t> _r_t(r_t);
+    auto van = GenVandermondeMatrix<ring2k_t>(world_size, world_size - th);
+    // TODO optimize me: all random shares can be done by a mmut between van^T *
+    // r_shrs van^T is a n-t by n r_shrs is a n by dn_times matrix
+    pforeach(0, dn_times, [&](int64_t idx) {
+      GfmpMatrix<ring2k_t> s_t(1, world_size);
+      for (auto i = 0; i < world_size; ++i) {
+        NdArrayView<ring2k_t> _r_shrs(r_shrs[i]);
+        s_t(0, i) = Gfmp<ring2k_t>(_r_shrs[idx]);
+      }
+      auto ret_t = s_t * van;
+
+      for (auto i = 0; i < (world_size - th); ++i) {
+        _r_t[idx * (world_size - th) + i] = ret_t(0, i).data();
+      }
+    });
+    auto out = r_t.slice({0}, {numel}, {});
+    return out.as(makeType<AShrTy>(field));
+  });
+}
+
+NdArrayRef P2A::proc(KernelEvalContext* ctx, const NdArrayRef& in) const {
+  const auto field = in.eltype().as<Ring2k>()->field();
+  return in.as(makeType<AShrTy>(field));
+}
+
+NdArrayRef A2P::proc(KernelEvalContext* ctx, const NdArrayRef& in) const {
+  const auto field = in.eltype().as<GfmpTy>()->field();
+  auto* comm = ctx->getState<Communicator>();
+  // we choose rank 0 as the P_pking for reconstructing secrets
+  auto arrays = comm->gather(in, 0, "send to pking");
+  NdArrayRef out = ring_zeros(field, in.shape());
```

Comment:
no need to allocate memory
```suggestion
  NdArrayRef out;
```